### ISSUE TITLE
terraform not correctly updating to correct git reference after initial download failure

### ISSUE THREAD

**Issue by willthames** [hashicorp/terraform#18797]
### Terraform Version

```
Terraform v0.11.8
```

### Terraform Configuration Files

None

### Debug Output

N/a

### Crash Output
N/A

### Expected Behavior

Module used is correct git reference

### Actual Behavior

Master branch of the module is used

### Steps to Reproduce

1. Create a module with a remote git repository. Create a master branch with an empty variables.tf. Commit this and push.
2. Change to a new branch, update variables.tf to add a new variable
```
variable "new_var" { }
```
3. Commit this but *do not push yet*
4. In the main terraform repo, add a main.tf that references the new remote repository referencing the as-yet unpushed commit
```
module "test" {
  source = "git::https://github.com/willthames/terraform-module-bug-test.git?ref=fb7a34e9a46b96c851b2f549f147790132215eed"
  new_var = "hello"
}
```
5. `terraform init`
6. This will fail due to the unpushed commit, as expected
7. Push the commit
8. `terraform init`
9. This will fail with 
  ```
  Error: module "test": "new_var" is not a valid argument
  ```
  Looking in the .terraform/modules as referenced by modules.json, the module is still on the master branch rather than the git reference.

Basically, it seems like the process is:
* Download module
* Update modules.json
* Attempt to change to the reference, but fail, so leave at default HEAD
* Subsequent terraform init runs then just use the cached module, and don't update to the proper reference.

**Comment by KadenLNelson** [comment]
related https://github.com/hashicorp/terraform/issues/19975 ?

**Comment by mildwonkey** [comment]
Indeed, this is a duplicate. The resolution is to run `terraform get -update`. Thanks!

**Comment by willthames** [comment]
Not sure that older tickets should be closed as a duplicate of a newer issue, but this one hasn't affected me since 0.12 so I'm fine with it being closed
